alacritty_terminal/src/term/search.rs

```diff
@@ -0,0 +1,794 @@
+        // Check case info for smart case
+        let has_uppercase = search.chars().any(|c| c.is_uppercase());
+
+        // Create Regex DFAs for all search directions.
+        let mut builder = dense::Builder::new();
+        let builder = builder.case_insensitive(!has_uppercase);
+
+        let left_fdfa = builder.clone().reverse(true).build(search)?;
+        let left_rdfa = builder.clone().anchored(true).longest_match(true).build(search)?;
+
+        let right_fdfa = builder.clone().build(search)?;
+        let right_rdfa = builder.anchored(true).longest_match(true).reverse(true).build(search)?;
+
+        Ok(RegexSearch { right_fdfa, right_rdfa, left_fdfa, left_rdfa })
+    }
+}
+
+impl<T> Term<T> {
+    /// Enter terminal buffer search mode.
+    #[inline]
+    pub fn start_search(&mut self, search: &str) {
+        self.regex_search = RegexSearch::new(search).ok();
+        self.dirty = true;
+    }
+
+    /// Cancel active terminal buffer search.
+    #[inline]
+    pub fn cancel_search(&mut self) {
+        self.regex_search = None;
+        self.dirty = true;
+    }
+
+    /// Get next search match in the specified direction.
+    pub fn search_next(
+        &self,
+        mut origin: Point<usize>,
+        direction: Direction,
+        side: Side,
+        mut max_lines: Option<usize>,
+    ) -> Option<Match> {
+        origin = self.expand_wide(origin, direction);
+
+        max_lines = max_lines.filter(|max_lines| max_lines + 1 < self.total_lines());
+
+        match direction {
+            Direction::Right => self.next_match_right(origin, side, max_lines),
+            Direction::Left => self.next_match_left(origin, side, max_lines),
+        }
+    }
+
+    /// Find the next match to the right of the origin.
+    fn next_match_right(
+        &self,
+        origin: Point<usize>,
+        side: Side,
+        max_lines: Option<usize>,
+    ) -> Option<Match> {
+        // Skip origin itself to exclude it from the search results.
+        let origin = origin.add_absolute(self, Boundary::Wrap, 1);
+        let start = self.line_search_left(origin);
+        let mut end = start;
+
+        // Limit maximum number of lines searched.
+        let total_lines = self.total_lines();
+        end = match max_lines {
+            Some(max_lines) => {
+                let line = (start.line + total_lines - max_lines) % total_lines;
+                Point::new(line, self.num_cols() - 1)
+            },
+            _ => end.sub_absolute(self, Boundary::Wrap, 1),
+        };
+
+        let mut regex_iter = RegexIter::new(start, end, Direction::Right, &self).peekable();
+
+        // Check if there's any match at all.
+        let first_match = regex_iter.peek()?.clone();
+
+        let regex_match = regex_iter
+            .find(|regex_match| {
+                let match_point = Self::match_side(&regex_match, side);
+
+                // If the matche's point is beyond the origin, we're done.
```

Comment:
matched? I don't quite like that `matche's`, not sure that it's a word.

Reply:
It should be `match's`, thanks.